### PR TITLE
Improve performance of file scanner

### PR DESCRIPTION
**Pull request by icewind1991**
Provides about ~2.5x speedup during my local testing while scanning a large local folder.

For details about how this is achieved see the individual commits

### CODE REVIEW

**Comment by rullzer** [comment]
is the 3rdparty update intentional here? seems not related or?

File: lib/private/Files/Cache/Scanner.php [hunk 1]
```diff
@@ -123,7 +123,7 @@ protected function getData($path) {
 	 * @param string $file
 	 * @param int $reuseExisting
 	 * @param int $parentId
-	 * @param array | null $cacheData existing data in the cache for the file to be scanned
+	 * @param array | null | false $cacheData existing data in the cache for the file to be scanned
```
Comment (kesselb) [inline lib/private/Files/Cache/Scanner.php]:
```suggestion
	 * @param array|null|false $cacheData existing data in the cache for the file to be scanned
```

File: lib/private/Files/Storage/Local.php [hunk 2]
```diff
@@ -150,6 +151,54 @@ public function stat($path) {
 		return $statResult;
 	}
 
+	/**
+	 * @inheritdoc
+	 */
+	public function getMetaData($path) {
+		$fullPath = $this->getSourcePath($path);
+		$stat = @stat($fullPath);
```
Comment (kesselb) [inline lib/private/Files/Storage/Local.php]: :rocket: I guess using stat directly is faster than `getPermissions` but also a bit scary.

**Comment by icewind1991** [comment]
> is the 3rdparty update intentional here? seems not related or?

Was accidental, fixed now

**Review by skjnldsv — APPROVED** [review]

File: lib/private/Files/Storage/Local.php [hunk 3]
```diff
@@ -151,6 +152,54 @@ public function stat($path) {
 		return $statResult;
 	}
 
+	/**
+	 * @inheritdoc
```
Comment (skjnldsv) [inline lib/private/Files/Storage/Local.php:156]: Missing public declaration?

Comment (icewind1991) [reply lib/private/Files/Storage/Local.php:156]: ?

The method already exists in the public interface (previously it inherited the implementation from `Common`)

Comment (skjnldsv) [reply lib/private/Files/Storage/Local.php:156]: I looked but didn't saw it, all good then!

**Review by rullzer — APPROVED** [review]

**Pull request merged**